src/libs/SidebarUtils.js

```diff
@@ -0,0 +1,315 @@
+
+    const filteredReports = _.filter(reports, (report) => {
+        if (!report || !report.reportID) {
+            return false;
+        }
+
+        const isChatRoom = ReportUtils.isChatRoom(report);
+        const isDefaultRoom = ReportUtils.isDefaultRoom(report);
+        const isPolicyExpenseChat = ReportUtils.isPolicyExpenseChat(report);
+        const participants = report.participants || [];
+
+        // Skip this report if it has no participants and if it's not a type of report supported in the LHN
+        if (_.isEmpty(participants) && !isChatRoom && !isDefaultRoom && !isPolicyExpenseChat) {
+            return false;
+        }
+
+        const hasDraftComment = report.hasDraft || false;
+        const iouReport = report.iouReportID && iouReports && iouReports[`${ONYXKEYS.COLLECTION.REPORT_IOUS}${report.iouReportID}`];
+        const iouReportOwner = report.hasOutstandingIOU && iouReport
+            ? iouReport.ownerEmail
+            : '';
+
+        const reportContainsIOUDebt = iouReportOwner && iouReportOwner !== currentUserLogin;
+        const shouldFilterReportIfEmpty = report.lastMessageTimestamp === 0
+
+            // We make exceptions for defaultRooms and policyExpenseChats so we can immediately
+            // highlight them in the LHN when they are created and have no messsages yet. We do
+            // not give archived rooms this exception since they do not need to be higlihted.
+            && !(!ReportUtils.isArchivedRoom(report) && (isDefaultRoom || isPolicyExpenseChat));
+
+        const shouldFilterReportIfRead = hideReadReports && !ReportUtils.isUnread(report);
+        const shouldFilterReport = shouldFilterReportIfEmpty || shouldFilterReportIfRead;
+        if (report.reportID.toString() !== currentlyViewedReportID
+            && !report.isPinned
+            && !hasDraftComment
+            && shouldFilterReport
+            && !reportContainsIOUDebt) {
+            return false;
+        }
+
+        // We let Free Plan default rooms to be shown in the App - it's the one exception to the beta, otherwise do not show policy rooms in product
+        if (ReportUtils.isDefaultRoom(report) && !Permissions.canUseDefaultRooms(betas) && ReportUtils.getPolicyType(report, policies) !== CONST.POLICY.TYPE.FREE) {
+            return false;
+        }
+
+        if (ReportUtils.isUserCreatedPolicyRoom(report) && !Permissions.canUsePolicyRooms(betas)) {
+            return false;
+        }
+
+        if (isPolicyExpenseChat && !Permissions.canUsePolicyExpenseChat(betas)) {
+            return false;
+        }
+
+        return true;
+    });
+
+    let orderedReports = _.sortBy(filteredReports, sortByTimestampDescending ? 'lastMessageTimestamp' : 'reportName');
+
+    if (sortByTimestampDescending) {
+        orderedReports.reverse();
+    }
+
+    // Move the archived Rooms to the last
+    orderedReports = _.sortBy(orderedReports, report => ReportUtils.isArchivedRoom(report));
```

Comment:
Yeah, that's not a bad idea. Every time I see this code I have a bad feeling about what it's doing. I have some nagging idea in the back of my mind that it could be done better (both for efficiency and for clarity). I haven't had time to really dig into it. I think I'll create a followup GH. Would that be OK for you?

Reply:
yep that sounds reasonable to me. I wouldn't mind looking into it.

Reply:
Here you go! https://github.com/Expensify/Expensify/issues/229617